### PR TITLE
Show Optimo promos in the WSOJ

### PR DESCRIPTION
**Pull request by DarioR01**
Resolves #WSTEAM1-140

**Overall change:**
Given I'm on a CPS page 
AND the recommenders feed contains an Optimo article in the top 4 position
When I view the CPS page 
Then the recommenders should show promos for both article types.

**Code changes:**

- Created a new utility function with signature `extractPromoData({promo})`. This function will pre-process Optimo and cps promos and creates a normalised data block used in `cpsRecommendations`.
- Minor adjustments to `cpsRecommendation` and `cpsRecommendationsImage` to adapt to the new normalised block.

---

- [ ] I have assigned myself to this PR and the corresponding issues
- [ ] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [ ] I have assigned this PR to the Simorgh project
- [ ] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: src/app/legacy/containers/CpsRecommendations/RecommendationsPromo/index.jsx [hunk 1]
```diff
@@ -103,9 +104,29 @@ const StyledHeadline = styled.div`
...
+        'text',
+      ],
+      promo,
+    );
+  const url =
+    pathOr(null, ['locators', 'assetUri'], promo) ||
+    pathOr(null, ['locators', 'canonicalUrl'], promo);
+  const indexImage =
+    pathOr(null, ['indexImage'], promo) ||
+    optimoToCPSImage({ cpsImage: pathOr(null, ['images'], promo) });
```
Comment (andrewscfc) [inline src/app/legacy/containers/CpsRecommendations/RecommendationsPromo/index.jsx]: This code here isn't explicit as to its intent, i.e. it processes the data differently depending on if it is a CPS recommendation or an Optimo recommendation. To be slightly more explicit and practice seperation of concerns you could create a new utility:
`const { headline, url, indexImage } = extractPromoData(promo);`

Inside that utility you could make it clear which logic is for CPS and which for Optimo e.g. 
```javascript
if(promo.locators.cpsUrn) {
  return {
     headline: pathOr(null, ['headlines', 'headline'], promo),
     ...
}

//optimo
return {
  headline: 
        pathOr(
      null,
      [
        'headlines',
        'promoHeadline',
        'blocks',
        0,
        'model',
        'blocks',
        0,
        'model',
        'text',
      ],
      promo,
    );
    ...
 }
 ```
 
 That way the `RecommendationsPromo` component itself doesn't need to know there is a difference between data formats for promos just that there is a seperate component that takes care of it. This is practicing the Open-Closed principle: https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle

File: src/app/legacy/containers/CpsRecommendations/RecommendationsPromoImage/index.jsx [hunk 2]
```diff
@@ -13,11 +13,19 @@ const RecommendationsImage = ({ indexImage, lazyLoad }) => {
...
+    altText,
+    copyrightHolder,
+    optimoLocator,
+    optimoOriginCode,
+  } = indexImage;
 
   const ratio = (height / width) * 100;
-  const originCode = getOriginCode(path);
-  const locator = getLocator(path);
+  const originCode = getOriginCode(path) || optimoOriginCode;
```
Comment (andrewscfc) [inline src/app/legacy/containers/CpsRecommendations/RecommendationsPromoImage/index.jsx]:
```suggestion
  const originCode = optimoOriginCode || getOriginCode(path);
```
Should this be the other way around? What does `getOriginCode` return in the event of an optimo `path` being provided?

This logic could also be encapsulated in the utility I suggested above, again so `RecommendationsPromoImage` is unware of the multiple types of promo data we need to support.

**Review by andrewscfc — COMMENTED** [review]
Few suggestions to meet some engineer principles that are good to follow, give me a shout if you wanna discuss!

File: src/app/legacy/containers/CpsRecommendations/RecommendationsPromo/index.jsx (hunk 1 above)
Comment (DarioR01) [reply src/app/legacy/containers/CpsRecommendations/RecommendationsPromo/index.jsx]: 👍, When I got some time to spare I need to revise some of these fundamentals, and actually practice how to apply them and recognise any violations. 
This is now fixed

File: src/app/legacy/containers/CpsRecommendations/RecommendationsPromoImage/index.jsx (hunk 2 above)
Comment (DarioR01) [reply src/app/legacy/containers/CpsRecommendations/RecommendationsPromoImage/index.jsx]: I believe the path would be null in this case and getOriginCode will return null. As you suggested this logic is now encapsulated in the `extractPromoData({promo})`, so it does not matter anymore.

**Review by amoore108 — APPROVED** [review]
Nice job! 👍🏼

**Review by andrewscfc — APPROVED** [review]
Thanks for applying my suggestions @DarioR01 ! If you want me to take you through some of the principles, very happy to give you the time. They are so helpful day to day when you know where to spot them 🙂

**Pull request merged**